### ISSUE TITLE
ejabberd_auth_http

### ISSUE THREAD

**Issue by melaniebalaz** [processone/ejabberd-contrib#200]
Hey there,

I am trying to get the ejabberd_auth_http module to work with the newest ejabberd version (the 17.03 beta one) just for testing purposes, and I have run into the following error:

> Application ejabberd exited with reason: no match of right hand value {error,{{'EXIT',{undef,[{cuesport,start_link,['ejabberd_auth_http_ip-IPADRESS',10,[fusco],{fusco,start_link,["https://IPAdressOfRemoteServer:443",[]]}],[]},{supervisor,do_start_child,2,[{file,"supervisor.erl"},{line,358}]},{supervisor,handle_start_child,2,[{file,"supervisor.erl"},{line,717}]},{supervisor,handle_call,3,[{file,"supervisor.erl"},{line,415}]},{gen_server,try_handle_call,4,[{file,"gen_server.erl"},{line,629}]},{gen_server,handle_msg,5,[{file,"gen_server.erl"},{line,661}]},...]}},...}} in ejabberd_auth_http:start/1 line 57 

I installed the module via the ejabberdctl module_install command and my config looks as such:

```
auth_method: http
auth_opts:
  host: "https://IPAdressOfRemoteServer:443"
  connection_pool_size: 10
  connection_opts: []
  basic_auth: ""
  path_prefix: "/drupal/auth/"
```

(obviously the IPAdressOfRemoteServer being the actual IP Adress)

Whenever I try to start the server with this config in, I get the before posted error in my log and when I check the status of ejabberd I get a:

> The node ejabberd@localhost is started with status: started
> ejabberd is not running in that node

Is this plugin just not compatible with the newest Ejabberd version? And if yes, which version is recommended to run the plugin with? I tried it on the 16.01 version before and there I got a whole other set of  

> call to undefined function ejabberd_auth_http:check_password

errors, when I attempted a test authentication (again module was installed via the ejabberdctl module_install tool).

Thank you!

**Comment by badlop** [comment]
> {undef,[{cuesport,start_link,

ejabberd_auth_http requires two erlang libraries called cuesport and fusco. They are included in the module source code, compiled, but some bug in ejabberd does not install them.

So, go to
> $HOME/.ejabberd-modules/sources/ejabberd-contrib/ejabberd_auth_http/deps/...

and copy all the *.beam and *.so files you can find in those subdirs. Copy them to:
> $HOME/.ejabberd-modules/ejabberd_auth_http/ebin/

Then restart ejabberd. If it throws some other error, please show them here... let's see if we can make the module to work.

**Comment by melaniebalaz** [comment]
Alright so!
First of all thank you for the super quick reply!

I did as instructed, and I got a bit further than before (I also switched over to a 16.12 version of the server, because I got the same error there and this will be the one I will continue working with after this module is working).
So in the /.ejabberd-modules/ejabberd_auth_http/ebin folder I added the following files I found in the two libraries and now the content looks like such:

> cuesport.beam  ejabberd_auth_http.beam  fusco.beam  fusco_binary.beam  fusco_cp.beam  fusco_drv.so  fusco_lib.beam  fusco_protocol.beam  fusco_sock.beam  scram2.beam

I wasnt sure whether it was important to uphold any sort of directory structure, as the beam and so files were in different folders, but as I understood it from your instructions it did not matter, but correct me if I am wrong and I should have moved the directories as well.

I was able to start the server after making those changes! Now I tried to connect to the server with Adium and I checked the logs and I could see that the connection was established and the authentication process was happening. However I got an error at this point, with the program crashing again. The error looks like the one I got before but it is on a different method so I am guessing we got a bit further than before. The logs look as such:

> 2017-03-13 08:53:50.894 [info] <0.624.0>@ejabberd_listener : accept:333 (#Port<0.17332>) Accepted connection ClientIP:50223 -> ServerIP:5222
> 2017-03-13 08:53:51.241 [error] <0.689.0> gen_fsm <0.689.0> in state wait_for_sasl_response terminated with reason: no match of right hand value {error,econnrefused} in ejabberd_auth_http:make_req/5 line 254
> 2017-03-13 08:53:51.241 [error] <0.689.0> CRASH REPORT Process <0.689.0> with 0 neighbours exited with reason: no match of right hand value {error,econnrefused} in ejabberd_auth_http:make_req/5 line 254 in p1_fsm:terminate/8 line 760

If we get this to work I can make a short article/post about what to do exactly to get the module to work for any others who try in the future.

Thank you for your help!

**Comment by melaniebalaz** [comment]
Actually, I think I got it to work now and I just had a config error on the 16.12 version! 

I checked the config again, and on this server in the auth http config part I hadnt entered the IP Adress yet of the server I actually want to authenticate against but still had localhost. I changed it to the actual server and checked the logs and it definitely looked like it sent out a request and tried to authenticate and I did not get any program crashes or any new erros in the error log.

So it seems just moving the libraries did the job, I will test out further!
Thanks for your help, I will let you know if I stumble upon any more errors. :)
